### PR TITLE
[MI-3206] Add script to create users and store user data

### PR DESCRIPTION
**Pull request by ayusht2810**
1. Added script to create users and store their data
2. Added script to clear store data
3. Added makefile

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -1,8 +1,31 @@
...
-	scripts.LoadConfig()
+	config, err := scripts.LoadConfig()
+	if err != nil {
+		panic(err)
+	}
+
+	args := os.Args
+	if len(args) > 1 {
+		switch args[1] {
+		case "create_users":
```
Comment (Kshitij-Katiyar) [inline main.go]: constants for these

File: serializers/config.go [hunk 2]
```diff
@@ -2,29 +2,29 @@ package serializers
...
 
 type ChannelsConfiguration struct {
-	DisplayName string 
-	Name string 
-	Type string
-	TeamId string
+	DisplayName string
+	Name        string
+	Type        string
+	TeamId      string
```
Comment (raghavaggarwal2308) [inline serializers/config.go]:
```suggestion
	TeamID      string
```

Comment (raghavaggarwal2308) [reply serializers/config.go]: check everywhere

File: serializers/response.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package serializers
+
+type UserResponse struct {
+	Id       string `json:"id"`
```
Comment (raghavaggarwal2308) [inline serializers/response.go]: same

File: scripts/create_users.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	userMap := make(map[string]interface{})
+	userMap[constants.NewUsersKey] = newUsers
+	userMapBytes, err := json.Marshal(userMap)
+	if err != nil {
+		return err
+	}
+
+	ioutil.WriteFile(constants.TempStoreFile, userMapBytes, 0644)
```
Comment (raghavaggarwal2308) [inline scripts/create_users.go]: Just a question, what is the use of 0644? Also, should we create a constant for this?

File: scripts/create_users.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+func CreateUsers(config *serializers.Config) error {
+	client := model.NewAPIv4Client(config.ConnectionConfiguration.ServerURL)
+	var newUsers []*serializers.UserResponse
+	for _, user := range config.UsersConfiguration {
+		createdUser, cErr := client.CreateUser(&model.User{
+			Username: user.Username,
+			Email:    user.Email,
+			Password: user.Password,
+		})
+
```
Comment (raghavaggarwal2308) [inline scripts/create_users.go]:
```suggestion
```

File: scripts/create_users.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+		if cErr.Error != nil {
+			return errors.New(cErr.Error.Message)
```
Comment (raghavaggarwal2308) [inline scripts/create_users.go]: can we use cErr.Error() here?

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (ayusht2810) [reply main.go]: I don't think these will be used anywhere else. We can create a constant later if needed.

File: scripts/create_users.go (hunk 6 above)
Comment (ayusht2810) [reply scripts/create_users.go]: no the type of `cErr.Error` is of the `*model.AppError` type

File: scripts/create_users.go (hunk 5 above)
Comment (ayusht2810) [reply scripts/create_users.go]: I think it is fine already

**Review by raghavaggarwal2308 — APPROVED** [review]

File: .gitignore [hunk 7]
```diff
@@ -1 +1,2 @@
 config/config.json
+temp_store/temp_store.json
```
Comment (manojmalik20) [inline .gitignore]:
```suggestion
temp_store/*
```

File: constants/constants.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+// Data keys
+const (
+	NewUsersKey = "createdUsers"
+)
+
+// Miscellaneous constants
+const (
+	FilePermissions = 0644
```
Comment (manojmalik20) [inline constants/constants.go]: Try to use these constants:
https://pkg.go.dev/os#pkg-constants

File: go.mod [hunk 9]
```diff
@@ -1,3 +1,48 @@
 module github.com/Brightscout/mattermost-load-test-scripts
 
 go 1.18
+
+require github.com/mattermost/mattermost-server/v5 v5.39.3
```
Comment (manojmalik20) [inline go.mod]: Why are we using this specific version?

File: main.go [hunk 10]
```diff
@@ -1,8 +1,31 @@
...
+			err = scripts.CreateUsers(config)
+		case "clear_store":
+			err = scripts.ClearStore()
+		default:
+			fmt.Println("Invalid arguments")
```
Comment (manojmalik20) [inline main.go]: I think we should probably use a logging package. Let's discuss this once.

File: makefile [hunk 11]
```diff
@@ -0,0 +1,5 @@
+create_users:
+	go run main.go create_users
```
Comment (manojmalik20) [inline makefile]: Create one command for creating binary of this package and then these commands will use that binary with the command-line arguments.

File: scripts/create_users.go (hunk 6 above)
Comment (manojmalik20) [inline scripts/create_users.go]:
```suggestion
			return errors.New(cErr.Error())
```

File: main.go (hunk 1 above)
Comment (Kshitij-Katiyar) [reply main.go]: but it would be nice to have all the functionalaties like create_users etc at one place

File: main.go [hunk 12]
```diff
@@ -1,8 +1,41 @@
...
+	"os"
+
+	"github.com/Brightscout/mattermost-load-test-scripts/scripts"
+	"go.uber.org/zap"
+)
 
-// TODO: remove later, if not needed
 func main() {
-	scripts.LoadConfig()
+	logger, _ := zap.NewProduction()
```
Comment (Kshitij-Katiyar) [inline main.go]: Is this error then handle this

File: main.go [hunk 13]
```diff
@@ -1,8 +1,41 @@
...
+	defer logger.Sync()
+	config, err := scripts.LoadConfig()
+	if err != nil {
+		logger.Error("failed to load the config",
```
Comment (Kshitij-Katiyar) [inline main.go]: same line

File: main.go [hunk 14]
```diff
@@ -1,8 +1,41 @@
...
+		case "clear_store":
+			err = scripts.ClearStore()
+		default:
+			err = errors.New("invalid arguments")
+		}
+	}
+
+	if err != nil {
+		logger.Error("failed to run the script",
+			zap.String("arg", args[1]),
```
Comment (Kshitij-Katiyar) [inline main.go]: same

File: constants/constants.go [hunk 15]
```diff
@@ -0,0 +1,18 @@
+package constants
+
+// File locations
+const (
+	ConfigFile    = "config/config.json"
+	TempStoreFile = "temp_store.json"
+)
+
+// Scripts arguments
```
Comment (manojmalik20) [inline constants/constants.go]:
```suggestion
// Script arguments
```

File: constants/constants.go [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+const (
+	CreateUsers = "create_users"
+	ClearStore  = "clear_store"
+)
+
+// Data keys
+const (
+	NewUsersKey = "createdUsers"
```
Comment (Kshitij-Katiyar) [inline constants/constants.go:17]:
```suggestion
	CreateUserKey  = "createdUsers"
```

File: main.go [hunk 17]
```diff
@@ -1,8 +1,40 @@
...
 
-// TODO: remove later, if not needed
 func main() {
-	scripts.LoadConfig()
+	logger, err := zap.NewProduction()
+	if err != nil {
+		panic(err)
+	}
+
+	defer logger.Sync()
```
Comment (Kshitij-Katiyar) [inline main.go:18]:
```suggestion
	defer logger.Sync()
	
```

File: main.go [hunk 18]
```diff
@@ -1,8 +1,40 @@
...
+		switch args[1] {
+		case constants.CreateUsers:
+			err = scripts.CreateUsers(config, logger)
+		case constants.ClearStore:
+			err = scripts.ClearStore()
+		default:
+			err = errors.New("invalid arguments")
+		}
+	}
+
```
Comment (Kshitij-Katiyar) [inline main.go]:
```suggestion
```

File: scripts/create_users.go (hunk 5 above)
Comment (Kshitij-Katiyar) [reply scripts/create_users.go]: fix this

File: makefile [hunk 19]
```diff
@@ -0,0 +1,10 @@
+GO ?= $(shell command -v go 2> /dev/null)
+
```
Comment (manojmalik20) [inline makefile]: Rename the file to `Makefile`.

**Review by manojmalik20 — APPROVED** [review]

File: constants/constants.go (hunk 16 above)
Comment (ayusht2810) [reply constants/constants.go:17]: this is removed in next PR

**Review by Kshitij-Katiyar — APPROVED** [review]

**Pull request merged**